# CSE 5243 - Introduction to Data Mining
## Homework 2: Classification

- Semester: Spring 2026
- Instructor: Tom Bihari
- Section Wed/Fri 12:45PM 
- Student Name: Nicholas McCracken
- Student Email: mccracken.161@osu.edu
***

***
### Use of Artificial Intelligence and Peer Collaboration in this Assignment
This course's syllabus incorporates standard syllabus statements maintained by the university.  See the statements regarding use of AI, Copyrights, etc., here:
- https://ugeducation.osu.edu/academics/standard-syllabus/standard-syllabus-statements
- https://ugeducation.osu.edu/academics/standard-syllabus/optional-syllabus-statements

Real-world development of analytics and engineering often requires significant collaboration and idea sharing. Collaborating with human peers and AI tools, including Generative AI, can be used to gather ideas, but assignment content and code must never be directly used from the peer or tool and then represented as your own original work. **The overall guidance follows the general Academic Integrity Policy: Your submission must be "your own original work" - not written by another person or AI.**

For this assignment:
- You may access online resources to view **examples** of the kinds of analyses you wish to perform in this asignment.
- You may search (or ask GenAI) for **examples** of short Python code snippets that show how to call a particular function, draw a particular kind of plot, etc.
- You **may not** ask GenAI to generate code specifically for this assignment, or for the given dataset.
- You **may not** copy/paste significant code snippets, etc., from online resources and claim it as your own.
- Similarly, your commentary, explanations, etc. in your assignment submission must be written by you alone, **not** assisted by GenAI or copied from another source.
- In all cases, you must comply with all **intellectual property** (e.g., copyright) requirements.

Given the rapid advances in GenAI, etc., it is difficult to provide a complete delineation of the specific boundaries of use. **If you have concerns or would like additional guidance, please reach out to me.**

In the following cell, **briefly describe** how you approached the use of the aforementioned resources (GenAI, etc.) for this assignment, and provide any thoughts or suggestions you have on the topic.  **This cell is not graded for specific content, but it must be filled in - the goal is to help us all build a general understanding of this landscape.**

I used GenAI very little on this assignment. It was helpful for some syntax questions, but I needed to go to the scikit learn documentation for information about the models. I think it can be a nice supplement to googling, especially if some information is difficult to find. In this assignment though, all the information I needed as readily available.

***
# Section: Overview

### Assignment Overview
This assignment covers the **steps 4 and 5 of the six steps** of the **CRISP-DM process model** (Modeling and Evaluation). (See the CRISP-DM materials on CARMEN.)

The **objectives** of this assignment are:
- Solve a business problem by creating, evaluating, and comparing three classification models, and produce the outputs needed to provide business value for your stakeholders.
- Experiment with built-in classification models in **scikit-learn**.

### Dataset
In this assignment, you will use the same dataset that was used for Homework 1 - an ALTERED copy of the “Hotel Booking Cancelations dataset.
- This dataset was pulled on 1/19/26 from: https://www.kaggle.com/datasets/muhammaddawood42/hotel-booking-cancelations
- For Homework 1, the dataset was **modified** to highlight important aspects of data mining (e.g., data cleaning).  However, the original "clean" version of the data has also been included for this assignment.
- **NOTE:** Since you already have pre-processed this dataset in the previous assignment, you may choose to use the "clean" dataset provided to you in this assignment, or your "cleaned up" dataset from the previous assignment, or you may re-do the work here.
- **DO NOT publish or otherwise use this dataset for purposes other than this assignment**.

### Problem Statement
Assume that you are the Director of Data Science for Buckeye Resorts, Inc. (BRI), an international hotel chain.
- As is the case for all hotel chains, reservation cancellations cause significant impacts to BRI, in profitability, logistics, and other areas.  The cost to BRI of a cancelled reservation is **\\$500** on average. 
- BRI wants to improve (decrease) the cancellation rates at its hotels, using more tailored interventions, based on newly available detailed data.  BRI processes **100,000** reservations per year, so an incremental improvement in cancellation rates would have a significant impact.
- The current **"OLD" intervention** is to charge a **\\$100** cancellation fee for all cancelled reservations.  However, this still results in a net **\\$400** loss (**\\$500 - \\$100**) for BRI for each cancellation, and causes negative customer impressions.  (For simplicity, assume this fee does not change the customers' likelihood of cancelling, and the given dataset reflects the "baseline" cancellation rate.)
- The specific **"NEW" intervention** is not yet known, but the assumption is that it would be based on predicting which reservations would be at risk of being cancelled and encouraging those customers to not cancel. If this **NEW intervention** is predicted to be **better** than the current **OLD intervention**, then it would be adopted.
  - One intervention being considered is to offer a special financial incentive to customers who have reservations, but who are “at risk” of cancellation.  BRI has performed a small pilot test, and has found that offering a **$20** gift card to a customer who is planning to cancel is effective **35%** of the time in inducing the customer not to cancel (NOTE: The customer keeps the gift card and may use it anywhere, regardless of whether they cancel or not. Yes, this is odd!).
- BRI leadership has asked your team to analyze the new data, and determine if it is suitable for developing analyses and models that would be effective in predicting which future reservations are likely to be at risk of cancellation, so the new intervention could be offered.

### Things To Do
**In the previous assignment**, you completed the sections for the first three steps of CRISP-DM.  You **explored** the dataset, and **prepared** a clean dataset from it that contains the kind of information you think might be useful.  You now will make use of the dataset.

You now must **develop** and **evaluate** specific models for predicting the cancellations.  You will try the **off-the-shelf KNN classifier**, and **two other classifiers of your choice**.

Some intial guidance / sugggestions:

- You must develop a cost model from the problem statement above.  Consider creating a table that lists the benefit and cost dollar amounts for a decision on a **single reservation**.  Note that the incentive will be "offered" if Predicted is True, and the incentive is "needed" if Actual is True:

| Actual "At Risk" | Predicted "At Risk" | Incentive Benefit | Incentive Cost | Net Benefit (Benefit-Cost) |
|---|---|---|---|---|
| False | False | 0 | 0 | 0 |
| False | True  | 0 | 20 | -20 |
| True  | False | 0 | 500 | -500 |
| True  | True  | 175 | 20 | 155 |

- Much of the code below may be repetitive.  Consider creating a few reusable functions that can be called for each of the models you build (e.g., evaluation functions).
- When evaluating your classification models, consider both the cost "per reservation" and "per year" overall (since the latter is important to the BOB decision-making process).

- **Follow the instructions** in each of the sections below.

It is essential that you **communicate** your goals, thought process, actions, results, and conclusions to the **audience** that will consume this work.  It is **not enough** to show just the code.  It is not appropriate to show long sections of **unexplained printout**, etc.  Be kind to your readers and provide value to them!

**ALWAYS follow this pattern** when doing **each portion** of the work.  This allows us to give feedback and assign scores, and to give partial credit.  Make it easy for the reader to understand your work.
- Say (briefly) **what** you are trying to do, and **why**.
- Do it (code).
- Show or describe the **result** clearly (and briefly as needed), and explain the significant **conclusions or insights** derived from the results. 

### Collaboration
For this assignment, you should work as an individual. You may informally discuss ideas with classmates, but your work should be your own.

### What you need to turn in
1) Code

- For this homework, the code is the Jupyter Notebook.  Use the provided Jupyter Notebook template, and fill in the appropriate information.
- You may use common Python libraries for I/O, data manipulation, data visualization, etc. (e.g., NumPy, Pandas, MatPlotLib,…) 
- You may not use library operations that perform, in effect, the “core” computations for this homework (e.g., If the assignment is to write a K-Means algorithm, you may not use a library operation that, in effect, does the core work needed to implement a K-Means algorithm.).  When in doubt, ask the grader or instructor.  (**Note: For this assignment, you *will* be using build in library functions, so you are permitted to do so.  You may not, however, make use of a single function that does *all* of the work for you.**
- The code must be written by you, and any significant code snips you found on the Internet and used to understand how to do your coding for the core functionality must be attributed.  (You do not need to attribute basic functionality – matrix operations, IO, etc.)
- The code must be commented sufficiently to allow a reader to understand the algorithm without reading the actual Python, step by step.
- **When in doubt, ask the grader or instructor.**

2) Written Report
- For this homework, the report is the Jupyter Notebook.  The report should be well-written.  Please proof-read and remove spelling and grammar errors and typos.
- The report should discuss your analysis and observations. Key points and findings must be written in a style suitable for consumption by non-experts.  Present charts and graphs to support your observations. If you performed any data processing, cleaning, etc., please discuss it within the report.

### Grading

1. Overall readability and organization of your report (5%)
- Is it well organized and does the presentation flow in a logical manner?
- Are there no grammar and spelling mistakes?
- Do the charts/graphs relate to the text?
- Are the summarized key points and findings understandable by non-experts?
- Do the Overview and Conclusions provide context for the entire exercise?
2. Evaluation Method (10%)
- Does your evaluation method meet the needs of the developer (you) as well as the needs of your business stakeholders?
- Is the evaluation method sound?
- Did you describe both the method itself and why you chose it?
3. Pre-Processing of the Dataset (5%)
- Did you make reasonable choices for pre-processing, and explain why you made them?
4. Evaluation of the OLD Intervention (5%)
- Are the calculations correct with respect to the problem statement?
4. Evaluation of the KNN Classifier (20%)
- Are your design choices, data pre-processing and coding correct?
- Is it well documented?
- Have you made an effort to tune it for good performance?
- Is the evaluation sound?
5. Evaluation of the Second Classifier (20%)
- Are your design choices, data pre-processing and coding correct?
- Is it well documented?
- Have you made an effort to tune it for good performance?
- Is the evaluation sound?
6. Evaluation of the Third Classifier (20%)
- Is your reasoning and coding correct?
- Is it well documented?
- Have you made an effort to tune it for good performance?
- Is the evaluation sound?
7. Comparison of the Three Classifiers (10%)
- Is the comparison sound?
- Did you choose a specific classifier as best and explain why?
8. Conclusions (5%)
- Did you summarize appropriately your critical findings. 
- Did you provide appropriate conclusions and next steps.

### How to turn in your work on Carmen

Submit to Carmen the Jupyter Notebook. You do not need to include the input data.

**HAVE FUN!**
***

***
# Section: Overview
- Insert a short description of the scope of this exercise, any supporting information, etc.
***

This exercise is focused on steps 4 and 5 (modeling and evaluation) of the CRISP-DM process using a hotel reservation dataset to predict whether a reservation is likely to be cancelled. This builds off the previous exercise, which included steps 2 and 3 of the CRISP-DM process for this same dataset and business problem. Cancellations cost BRI $500. The current intervention demands a $100 cancellation fee, which is still a net loss of $400 and is dissatisfying for customers. BRI is considering whether to implement a new intervention where customers at risk of cancellation receive a $20 gift card, which prevents cancellations 35% of the time. 

It is our goal to identify to evaluate whether this approach would reduce the losses incurred by BRI from cancellations. Three classification models will be developed: KNN, logistic regression, and a decision tree classifier. The models will be evaluated using non-cost and cost-based measures to determine which intervention provides the most financial value.

***
# Section: Setup
- Add any needed imports, helper functions, etc., here.
***

In [228]:
import numpy as np
import pandas as pd

#!pip install matplotlib
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline
#matplotlib.use('Qt5Agg')

import seaborn as sns

pd.set_option('display.max_columns', 50) #include to avoid ... in middle of display

In [229]:
def make_dataframe_from_Xy(X, Xcols, y, ycol):
    # Helper function to take individual X and y arrays and return a dataframe
    # EXAMPLE CALL: df = make_dataframe_from_Xy( X, ['age', 'height'], y, 'happy')
    df = pd.DataFrame(X, columns=Xcols)
    df[ycol] = y.tolist()
    return df

In [230]:
def make_Xy_from_dataframe(df, Xcols, ycol):
    # Helper function to take a dataframe and return individual X and y arrays
    # EXAMPLE CALL: X, y = make_Xy_from_dataframe(df, ['age', 'height'],'happy')
    X = df[Xcols].to_numpy()
    y = df[ycol].to_numpy()
    return X, y

***
# Section: 1 - Evaluation Method
- Define measures for evaluating the classification models you develop.  Explain why the measures you choose provide a useful view into the value and usefulness of the model you eventually chose for the company to use.  **Note: In this section, you should define and explain your measures.  You may create a reusable function here if you like.  You then will use the functions in later sections.**
- Define two types:
***

***
## Section: 1.1 - Define measures that **do not** include the cost information
- (e.g., confusion matrices, accuracy, precision, recall, F-measures, etc.).
- Consider using: from sklearn.metrics import classification_report, confusion_matrix
- Create a reusable function here, since you will need it multiple times below.
***

I use several classification measures from sklearn.metrics to evaluate how well each classifier predicts cancellations. The evaluate_classification function below calculates accuracy, prints a classification report (includes precision, recall, and F1-score), and shows the confusion matrix. 

The confusion matrix is returned so it can be reused when calculating measures with cost information.

In [231]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

def evaluate_classification(y_true, y_pred):
    print(f"Accuracy: {accuracy_score(y_true, y_pred):.4f}")

    print("CLASSIFICATION REPORT")
    print(classification_report(y_true, y_pred))
        
    cm = confusion_matrix(y_true, y_pred)

    print("\nCONFUSION MATRIX")
    print("               No Cancel | Cancel")
    print(f"Actual No Cancel   {cm[0][0]:6} | {cm[0][1]:6}")
    print(f"Actual Cancel      {cm[1][0]:6} | {cm[1][1]:6}")

    TN, FP, FN, TP = cm.ravel()

    print("\nCONFUSION MATRIX BREAKDOWN")
    print(f"True Negatives  (Correct No Cancel): {TN}")
    print(f"False Positives (Predicted Cancel but didn't): {FP}")
    print(f"False Negatives (Missed Cancel): {FN}")
    print(f"True Positives  (Correct Cancel): {TP}")
    print()
    
    return cm

Accuracy gives an overall measure for corectness, while the classification report shows how well the model idenitifes each class. Though, there are only two classes in this binary classification problem of "at risk" or "not at risk". The confusion matrix is vital for cost evaluation. These three measures give us more than enough context to evaluate a classifier.

***
## Section: 1.2 - Define measures that **do** include the cost information
- (e.g., using cost matrices).
- Create a reusable function here, since you will need it multiple times below.
    - For example, create a function that takes a confusion matrix and calculates the cost.
***

We need to use a cost matrix based on the problem statement to gleam significance from the prediction accuracy.

If we predict that a reservation is at risk of cancellation, BRI offers a $20 gift card. When a customer would actually cancel, this prevents cancellation 35% of the time. Since cancellation costs $500, the incentive benefit of correctly prediction a cancellation with this new incentive is 0.35 * $500 = $175. Subtracting the gift card cost, the net benefit becomes $155. The other cases are self explanatory.

| Actual "At Risk" | Predicted "At Risk" | Incentive Benefit | Incentive Cost | Net Benefit (Benefit-Cost) |
|---|---|---|---|---|
| False | False | 0 | 0 | 0 |
| False | True  | 0 | 20 | -20 |
| True  | False | 0 | 500 | -500 |
| True  | True  | 175 | 20 | 155 |

The cost matrix used in the calculate_cost function below is derived from the net benefit column of this table for each case.

In [232]:
def calculate_cost(cm):
    cost_matrix = [
        [0, -20], 
        [-500, 155]
    ]
    
    total_cost = 0
    for row in range(2):
        for col in range(2):
            total_cost += cost_matrix[row][col] * cm[row][col]
            
    return total_cost

By distilling the confusion matrix for a classifier into a total cost value, we can compare classifiers based on their true monetary impact, rather than only accuracy. This ties in the classifier performance with actual business need, and can allow BRI to make a decision on how to best reduce cancellation losses.

***
# Section: 2 - Pre-Processing of the Dataset
- NOTE: You have done much of the data preprocessing in the previous assignment, so **you don't have to re-do it** here.  You can either **copy the necessary code** from the previous assignment, or use the **clean dataset** from the previous assignment and load it here.
***

***
## Section: 2.1 - Load the dataset
- Load the dataset.
***

In [233]:
# Note: Edit the path here to match your setup.
#   The "tiny" dataset contains just a few records to make it quick to test code.
#   The "with_errors" dataset is the same as the one for the Homework 1 assignment.
#   The "clean" dataset is the original, full dataset, with no added errors.
#data_df = pd.read_csv('./data/hotel_booking_tiny_V1.csv')
#data_df = pd.read_csv('./data/hotel_booking_with_errors_V1.csv')
data_df = pd.read_csv('./data/hotel_booking_clean_V1.csv')

# Note: Do not use the ROW attribute as data, and do not include it in any analyses.  It's only purpose is to support referring to a row in the data for documentation purposes, if any.

In [234]:
data_df.head().T

,0,1,2,3,4
ROW,1,2,3,4,5
is_canceled,0,0,0,0,0
lead_time,342,737,7,13,14
arrival_date_year,2015,2015,2015,2015,2015
hotel,Resort Hotel,Resort Hotel,Resort Hotel,Resort Hotel,Resort Hotel
arrival_date_month,July,July,July,July,July
arrival_date_week_number,27,27,27,27,27
arrival_date_day_of_month,1,1,1,1,1
stays_in_weekend_nights,0,0,0,0,0
stays_in_week_nights,0,0,1,1,2


In [235]:
data_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 37 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   ROW                             119390 non-null  int64  
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   hotel                           119390 non-null  str    
 5   arrival_date_month              119390 non-null  str    
 6   arrival_date_week_number        119390 non-null  int64  
 7   arrival_date_day_of_month       119390 non-null  int64  
 8   stays_in_weekend_nights         119390 non-null  int64  
 9   stays_in_week_nights            119390 non-null  int64  
 10  adults                          119390 non-null  int64  
 11  children                        119386 non-null  float64
 12  babies                     

***
## Section: 2.2 - Explore the dataset
- As in Homework 1, explore the attributes briefly. Reference the website listed in the Introduction.
- Provide basic statistics for the attributes.
- List which attributes are Nominal (even though they are encoded as numbers), Ordinal, Interval, Ratio.
- **NOTE: Just summarize here.  You will need to know which attributes are Nominal, etc., so it would be useful to list them here.**
***

In [236]:
data_df.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
ROW,119390.0,NaN,NaN,NaN,59695.5,34465.068657,1.0,29848.25,59695.5,89542.75,119390.0
is_canceled,119390.0,NaN,NaN,NaN,0.370416,0.482918,0.0,0.0,0.0,1.0,1.0
lead_time,119390.0,NaN,NaN,NaN,104.011416,106.863097,0.0,18.0,69.0,160.0,737.0
arrival_date_year,119390.0,NaN,NaN,NaN,2016.156554,0.707476,2015.0,2016.0,2016.0,2017.0,2017.0
hotel,119390,2,City Hotel,79330,NaN,NaN,NaN,NaN,NaN,NaN,NaN
arrival_date_month,119390,12,August,13877,NaN,NaN,NaN,NaN,NaN,NaN,NaN
arrival_date_week_number,119390.0,NaN,NaN,NaN,27.165173,13.605138,1.0,16.0,28.0,38.0,53.0
arrival_date_day_of_month,119390.0,NaN,NaN,NaN,15.798241,8.780829,1.0,8.0,16.0,23.0,31.0
stays_in_weekend_nights,119390.0,NaN,NaN,NaN,0.927599,0.998613,0.0,0.0,1.0,2.0,19.0
stays_in_week_nights,119390.0,NaN,NaN,NaN,2.500302,1.908286,0.0,1.0,2.0,3.0,50.0


***
## Section: 2.3 - Revise the dataset
- **NOTE: To run each classifier below, you must have a dataset that is suitable for that classifier. This can be done here in addition to, or instead of, transformations done in the individual sections below.**
- Review the meanings of the attributes and **consider** removing redundant or (likely) irrelevant attributes, combining attributes, etc., to reduce the number of attributes.
  - NOTE: You can keep all attributes if you choose!
  - You may choose to use techniques such as those you used in Homework 1 to analyze the impacts of individual attributes on the CLASS attribute.
- Consider transforming the remaining attributes (e.g., using the data dictionary to replace the numbers with text values for some attributes – this might or might not be useful), normalizing / scaling values, encoding labels (if necessary), etc.
- If you want to do any additional transformations, you can do them here.
  - **Note: You also may need to do some specific transformations below for each of the classifiers you choose.**
- Optionally (you will need this below): Split the dataset into a Training dataset and a Test dataset based on the class attribute.  Keep them separate and use the Training dataset for training/tuning and the Test dataset for testing. For consistency, use the **train_test_split** operation available in SciKit Learn (use a specific random seed, so it is reproducible).
  - from sklearn.model_selection import train_test_split
  - X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
- Describe what you chose to do (and not do), and why.
- **IMPORTANT: Any transformations you do to the datasets (particularly the Test dataset) must not artificially impact the evaluation measures.  We want the chosen classification model to work "in the real world", and the Test dataset is an approximation of the real world.**
***

Before training the classifiers, we need to prep the dataset into features (X) and the target variable (y), which happens to be is_canceled in this dataset. Attributes that are of extremely high uniqueness (i.e. personally identifying information) are dropped from X because they provide no useful information to the classifier. Some missing values in the children and country columns also need to be filled to avoid processing errors.

Some of these attributes are categorical (i.e. hotel_type), they must be converted into a numerical form so they can be easily proccessed when training our classifier. Label encoding can be applied for this purpose.

The dataset can also be split into training and test sets. The training set is used to train the classifiers, while the test set evaluates them based on unseen data. A fixed random speed of 42 is used to ensure reproducible results.

In [237]:
from sklearn.calibration import LabelEncoder
from sklearn.model_selection import train_test_split

# Get target variable
y = data_df["is_canceled"]

# Drop unnecessary columns
X = data_df.drop(columns=[
    "ROW",
    "is_canceled",
    "name",
    "email",
    "phone-number",
    "credit_card",
    "reservation_status",
    "reservation_status_date",
    "agent",
    "company"
])

# Fill remaining missing values
X["children"] = X["children"].fillna(0)
X["country"] = X["country"].fillna("Unknown")

# Convert categorical variables to numeric using label encoding
for col in X.select_dtypes(include="object").columns:
    X[col] = LabelEncoder().fit_transform(X[col])

# 67% training, 33% testing
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.33, random_state=42
)

# Sanity check to ensure expected number of records in input sets
print(X_train.shape, X_test.shape)

(79991, 27) (39399, 27)


/var/folders/5q/blczwkcj4jz5t90lvz2rkp0h0000gn/T/ipykernel_80992/3102426517.py:26: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in X.select_dtypes(include="object").columns:


After preprocessing, all features have numericcal representations, so the classifiers can easily operate on the data. The datatset is now divided into a training and test set so the model can be evaluated on data not yet seen by the classifier. This can allow us to check for overfitting, so we can ensure the model is not simply memorizing patterns from the training data.

***
# Section: 3 - Evaluate the Impact of the OLD Intervention
- In order to determine if the NEW intervention is worthwhile, we need to calculate the impact of cancellations using the OLD (current) intervention.  This will allow us to evaluate whether the NEW intervention is "better".
- Assume that the given dataset is a good representation of the typical behaviors of the customers. Use the dataset and the information in the Problem Statement at the top of this notebook to calculate:
  - The average amount of money the company loses due to cancellations, per individual reservation, if there is NO intervention. (HINT: Consider the probability that an individual reservation is cancelled.)
  - The total amount of money the company loses due to cancellations in a typical year, if there is NO intervention.  (HINT: Consider the typical number of reservations per year.)
  - The average amount of money the company loses due to cancellations, per individual reservation, using the OLD intervention.
  - The total amount of money the company loses due to cancellations in a typical year, using the OLD intervention.
***

We need to establish a baseline with the current intervention before we can deccide whether the new intervention is an improvement. Using the target value, we can calculate the rate at which a reservation is cancelled. With this cancel_rate, we can estimate the financial loss per reservation both with no intervention and the current intervention. 

Without intervention, each cancellation costs BRI $500. With the current intervention of an $100 cancellation fee, each cancellation costs BRI $400. We'll use these to estimate average loss per reservation and each year, assuming BRI has 100,000 reservations per year.

In [238]:
cancel_rate = data_df["is_canceled"].mean()

print(f"Cancellation rate: {cancel_rate:.4f}")

# From problem statement
reservations_per_year = 100000
cancel_cost = 500
cancel_fee = 100

# NO INTERVENTION
loss_per_reservation_no = cancel_rate * cancel_cost
loss_per_year_no = loss_per_reservation_no * reservations_per_year

print("\nNO INTERVENTION")
print(f"Loss per reservation: {loss_per_reservation_no:.2f}")
print(f"Loss per year: {loss_per_year_no:.0f}")

# OLD INTERVENTION
loss_per_cancel_old = cancel_cost - cancel_fee
loss_per_reservation_old = cancel_rate * loss_per_cancel_old
loss_per_year_old = loss_per_reservation_old * reservations_per_year

print("\nOLD INTERVENTION")
print(f"Loss per reservation: {loss_per_reservation_old:.2f}")
print(f"Loss per year: {loss_per_year_old:.0f}")

Cancellation rate: 0.3704

NO INTERVENTION
Loss per reservation: 185.21
Loss per year: 18520814

OLD INTERVENTION
Loss per reservation: 148.17
Loss per year: 14816651


From the given datatset, approximately 37% of reservations are cancelled. Without intervention, there's an average loss is $185.21 per reservation, which is about $18.52 million per year.

With the current intervention of the $100 cancellation fee, the average loss $148.17 per reservation, a $37.04 improvement from no intervention, and about $14.82 million per year. 

To determine whether the new intervention is better at reducing losses, we can compare it's average loss per reservation and loss per year to the current intervention in proceeding steps.

***
# Section: 4 - Evaluation of the Off-The-Shelf KNN Classifier
- Select the KNN classifier from the SciKit Learn library and run it on the dataset.
***

***
## Section: 4.1 - Configure the off-the-shelf KNN classifier
- Use the KNeighborsClassifier from the SciKit Learn library
- Explain all setup, parameters and execution options you chose to set, and why.
***

The first classifier is a KNN classifier from the SciKit Learn library. KNN is a distance based algorithm that predicts the class of a data point by examining the K MMost similiar data points in the training datatset, and assigning the majority class among said neighbors. KNN can capture patterns of reservations with similiar features. The primary parameter for the classifier is K, which determines how many neighbors are considered in each prediction. Seven different K values are tested and compared in the following sections.

In [239]:
from sklearn.neighbors import KNeighborsClassifier

Using the Scikit Learn KNeighborsClassifier is pretty straightforward. All the default configurations will serve for our purposes here.

***
## Section: 4.2 - Revise the dataset specifically for the chosen classifier
- Depending on the classifier you have chosen (KNN in this case), you may need / want to revise the dataset specifically for the charactistics of the classifier here.  For example, KNN is generally used for Numerical (not Categorical) attributes.  So it might be useful to encode the categorical attributes into numerical attributes for use here.
- This can be done here in addition to, or instead of, transformations done in Section 2.3 above.
- Explain what you have chosen to do, or not do, and why.
***

Since the dataset was already claned and categorical variables were converted to numeric during preprocessing in Section 2.3, the only additional preparation needed for the KNN is to normalize the features using StandardScaler. This ensures that all features have a mean of zero and std of one, such that all features will contribute more equally to the distance calculation sused in the KNN algorithm.

In [240]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

After scaling, the dataset is ready to be used by a distance based classifier like our KNN. This should make the KNN more reliable, and ensure that no single feature overly-sways the distance calculation.

***
## Section: 4.3 - Run and evaluate the classifier
- Try several values of the K parameter and compare the results.
- Evaluate the performance of the classifier, using the evaluation methods you defined above.
***

The only hyperparameter used in the KNN is K, which determines how many neighboring data points are considered when classifying a point. Larger K values make the classification broader and smoother, which could potentially mean more stable predictions with reduced overfitting. 

To find the balance between sensitivity to individual data points while mitigating overfitting, we evaluate seven different K values using the non-cost and cost-based evaluation functions defined in section 1.

In [241]:
k_values = [3, 5, 7, 10, 25, 50, 100]

for k in k_values:
    print(f"K = {k}")
    
    knn = KNeighborsClassifier(n_neighbors=k)
    
    # Train model
    knn.fit(X_train_scaled, y_train)
    
    # Make predictions
    y_prediction = knn.predict(X_test_scaled)
    
    # Evaluate classifier
    cm = evaluate_classification(y_test, y_prediction)
    
    # Evaluate financial impact
    total_cost = calculate_cost(cm)
    
    print("Total Cost:", total_cost)

K = 3
Accuracy: 0.8268
CLASSIFICATION REPORT
              precision    recall  f1-score   support

           0       0.86      0.87      0.86     24754
           1       0.77      0.76      0.76     14645

    accuracy                           0.83     39399
   macro avg       0.82      0.81      0.81     39399
weighted avg       0.83      0.83      0.83     39399


CONFUSION MATRIX
               No Cancel | Cancel
Actual No Cancel    21507 |   3247
Actual Cancel        3576 |  11069

CONFUSION MATRIX BREAKDOWN
True Negatives  (Correct No Cancel): 21507
False Positives (Predicted Cancel but didn't): 3247
False Negatives (Missed Cancel): 3576
True Positives  (Correct Cancel): 11069

Total Cost: -137245
K = 5
Accuracy: 0.8298
CLASSIFICATION REPORT
              precision    recall  f1-score   support

           0       0.86      0.88      0.87     24754
           1       0.78      0.75      0.77     14645

    accuracy                           0.83     39399
   macro avg       0.

The KNN classifier produced largely consistent accuracy across smaller values of K. The accuracy was around 83% for K=3, 5, and 7. The highest accuracy was =10 at 0.8311, but this improvement is minimal over the smaller K values. 

Looking at the confusion matricies, the classifier performs better at identifying reservations that do not cancel (class 0) than those that do cancel (class 1). The model becomes more conservative and general in cancellation predictions as K increases, as to be expected. This means fewer false positives, but much more false negatives.

When it comes to cost, the smaller K values are significantly better. The lowest cost is at K=3, while the larger K's are dramatically worse for BRI. Those false negatives missing cancellation risks is very costly for the company, which is why the higher K values perform poorly. Accuracy does not necessarily translate to best financial outcome.

***
## Section: 4.4 - Evaluate the choice of the KNN classifier
- What characteristics of the problem and data made KNN a good or bad choice?
***

KNN is a simple baseline classifier for prediction reservation cancellations. If reservations with similiar characteristics tend to behave similiar with respect to cancellations, this distance-based approach ccan be quite effective.

There are some limitations at the cost of this simplicity though. The algorithm can become computationally expensive for very large datasets, or datasets with lots of features. I tried a one-shot encoding before my label encoding solution in section 2.3, which caused the Jupyter kernel to crash when runnning KNN because of the large number of columns. Also these distance-based models are very sensitive to features that aren't super relevant. Some of the preprocessing steps which removed useless columns helps, but it's still a concern. In later sections, we'll see if alternate models can remedy some of these concerns.

***
# Section: 5 - Evaluation of Off-The-Shelf Classifier #2
- As with the KNN classifier above, choose another classifier from the SciKit Learn library (Decision Tree, SVM, Logistic Regression, etc.) and run it on the dataset.
***

***
## Section: 5.1 - Configure the classifier
- Use the appropriate classifier from the SciKit Learn library.
- Explain all setup, parameters and execution options you chose to set, and why.
***

The first classifier is a logistic regression classifier from the SciKit Learn library. Logistic regression predicts the probability that a data point belongs to a class using a logistic function which is applied to a linear combination of input features. This can be applied to our binary classification problem of cancelled or not cancelled. Logistic regression offers a much different approach from the distance based KNN in the previous section.

In [242]:
from sklearn.linear_model import LogisticRegression

Using the Scikit Learn Logistic Regression classifier is pretty straightforward. All the default configurations will serve for our purposes here; the off-the-shelf classifier is already efficient with large datasets.

***
## Section: 5.2 - Revise the dataset specifically for the chosen classifier
- Depending on the classifier you have chosen, you may need / want to revise the dataset specifically for the charactistics of the classifier here.
- This can be done here in addition to, or instead of, transformations done in Section 2.3 above.
- Explain what you have chosen to do, or not do, and why.
***

Logistic regression benefits from data standardization, as does the KNN classifier. That same work can be reused here.

In [243]:
X_train_lr = X_train_scaled
X_test_lr = X_test_scaled

Reusing the same scaled dataset across each model will ensure consistency, and guarentees a fair comparision between logistic regression and the other classifiers.

***
## Section: 5.3 - Run and evaluate the classifier
- Try several values of the parameters (if appropriate) and compare the results.
- Evaluate the performance of the classifier, using the evaluation methods you defined above.
***

The key hyperparameter used in logistic regression is C, which controls the strength of regularization that is applied during training. This regularization helps prevent overfitting by penalizing very large coefficients in the model. Smaller C values have stronger regularization and thus simpler models, while larger values allow for more room for close fitting to the training data.

To properly evaluate how regularization affects performance of the classifier, five C values are tested. 

In [244]:
C = [0.01, 0.1, 1, 10, 100]

for c in C:
    print(f"C = {c}")

    logistic_regression = LogisticRegression(C=c, max_iter=1000)

    # Train model
    logistic_regression.fit(X_train_lr, y_train)

    # Make predictions
    y_prediction = logistic_regression.predict(X_test_lr)

    # Evaluate classifier
    cm = evaluate_classification(y_test, y_prediction)

    # Evaluate financial impact
    total_cost = calculate_cost(cm)

    print("Total Cost:", total_cost)

C = 0.01
Accuracy: 0.7949
CLASSIFICATION REPORT
              precision    recall  f1-score   support

           0       0.79      0.92      0.85     24754
           1       0.81      0.58      0.68     14645

    accuracy                           0.79     39399
   macro avg       0.80      0.75      0.76     39399
weighted avg       0.80      0.79      0.79     39399


CONFUSION MATRIX
               No Cancel | Cancel
Actual No Cancel    22775 |   1979
Actual Cancel        6102 |   8543

CONFUSION MATRIX BREAKDOWN
True Negatives  (Correct No Cancel): 22775
False Positives (Predicted Cancel but didn't): 1979
False Negatives (Missed Cancel): 6102
True Positives  (Correct Cancel): 8543

Total Cost: -1766415
C = 0.1
Accuracy: 0.7952
CLASSIFICATION REPORT
              precision    recall  f1-score   support

           0       0.79      0.92      0.85     24754
           1       0.81      0.59      0.68     14645

    accuracy                           0.80     39399
   macro avg    

The results of the logistic regression classifier were very similiar across all tested C values. Accuracy remmained around 79.5% for each C, which indicates that the level of regularization didn't have much impact on overall performance of predictions. 

The confusion matricies show that the model, similiarily to the KNN, is good at identifying class 0 do not cancel reservations. It is less good at the do cancel class 1 reservations, since the recall value for those is lower. The model thus misses a significant number of actual cancellations for BRI.

Larger C values slightly reduce financial losses. C=100 produces the lowest total cost among any of the C values. Despite this, all logistic regression models produce much worse results than the best KNN configuration, likely due to the large number of false negatives in this classifier.

***
## Section: 5.4 - Evaluate the choice of the classifier
- What characteristics of the problem and data made the classifier a good or bad choice?
***

Logistic regression is computationally efficient and very easy to implement. Predictions can be broken down by how each feature influences cancellation likelihood, which is intuitive and helpful info. It is much more scaleable with large datasets than KNN. However, that's not very significant for the dataset in this problem.

Sincce logistic regression assumes a linear relationship between input features and cancellation probability, it fails to capture more complex/nonlinear relationships. It is lesser when it comes to pattern recognition compared to the KNN, as shown by the lower accuracy and worse finanicla outcome. The relationship between features and cancellations is more complex than this simple linear model can handle.

***
# Section: 6 - Evaluation of Off-The-Shelf Classifier #3
- As with the KNN classifier above, choose another classifier from the SciKit Learn library (Decision Tree, SVM, Logistic Regression, etc.) and run it on the dataset.
***

***
## Section: 6.1 - Configure the classifier
- Use the appropriate classifier from the SciKit Learn library.
- Explain all setup, parameters and execution options you chose to set, and why.
***

The third classifier is a decision tree classifier. Decision trees predict the class of a data point by learning a sequence of decision rules that split the dataset into classes based on feature values. These rules form a tree structure. Each internal node represents a decision on a feature, and each leaf node is a class prediction (cancelled 1 or not cancelled 0). 

In [245]:
from sklearn.tree import DecisionTreeClassifier

Using the Scikit Learn DecisionTreeClassifier is pretty straightforward. All the default configurations will serve for our purposes here.

***
## Section: 6.2 - Revise the dataset specifically for the chosen classifier
- Depending on the classifier you have chosen, you may need / want to revise the dataset specifically for the charactistics of the classifier here.
- This can be done here in addition to, or instead of, transformations done in Section 2.3 above.
- Explain what you have chosen to do, or not do, and why.
***

Unlike KNN and logistic regression, decision trees do not need standardization since they do not rely on distance calculations or linear relationships between features. Since feature scaling is not required, we can train this classifier on the original feature matricies from section 2.3.

In [246]:
X_train_dt = X_train
X_test_dt = X_test

Since decision trees do not care about feature scaling, no additional transformations are required.

***
## Section: 6.3 - Run and evaluate the classifier
- Try several values of the parameters (if appropriate) and compare the results.
- Evaluate the performance of the classifier, using the evaluation methods you defined above.
***

Perhaps the most important parameter of decision trees is max_depth, which limits how the deep the tree can grow. Smaller depth values produce simpler models, while larger depths capture more complexity but increase overfitting risk.

Several max_depth values are tested to determine how complexity of the model affects its performance.

In [247]:
max_depths = [3, 5, 10, 25, None]

for max_depth in max_depths:

    print(f"Max Depth = {max_depth}")

    dt = DecisionTreeClassifier(max_depth=max_depth, random_state=42)

    # Train model
    dt.fit(X_train_dt, y_train)

    # Make predictions
    y_pred = dt.predict(X_test_dt)

    # Evaluate classifier
    cm = evaluate_classification(y_test, y_pred)

    # Evaluate financial impact
    total_cost = calculate_cost(cm)

    print("Total Cost:", total_cost)

Max Depth = 3
Accuracy: 0.7505
CLASSIFICATION REPORT
              precision    recall  f1-score   support

           0       0.72      1.00      0.83     24754
           1       1.00      0.33      0.50     14645

    accuracy                           0.75     39399
   macro avg       0.86      0.66      0.67     39399
weighted avg       0.82      0.75      0.71     39399


CONFUSION MATRIX
               No Cancel | Cancel
Actual No Cancel    24733 |     21
Actual Cancel        9810 |   4835

CONFUSION MATRIX BREAKDOWN
True Negatives  (Correct No Cancel): 24733
False Positives (Predicted Cancel but didn't): 21
False Negatives (Missed Cancel): 9810
True Positives  (Correct Cancel): 4835

Total Cost: -4155995
Max Depth = 5
Accuracy: 0.7911
CLASSIFICATION REPORT
              precision    recall  f1-score   support

           0       0.78      0.93      0.85     24754
           1       0.83      0.55      0.66     14645

    accuracy                           0.79     39399
   macr

The result clearly show that increasing tree depth significantly improves classification performances and total cost outcome. 

Shallow trees such as max_depth=3 miss a large number of cancellations due to the large quantity of false negatives. As the maximmum depth increases, the decision tree classifier becomes better at identifying cancellations. Total cost positively correlates with accuracy in this classifier.

The no maximum depth tree has the best outcome out of any classifer configuration tested.

***
## Section: 6.4 - Evaluate the choice of the classifier
- What characteristics of the problem and data made the classifier a good or bad choice?
***

The decision tree classifier performs very well for this problem. Compared to previous models, the decision tree is more capable when it comes to capturing the complex relations between features and cancellation behavior. The results show that the deeper trees perform better for this dataset, suggesting the relationships are quite complex. The decision tree with no max depth has the best accuracy and financial outcome, without much overfitting.

***
# Section: 7 - Comparison of the Three Classifiers
***

***
## Section: 7.1 - Compare the performance of these classifiers to each other
- What are their strong and weak points?
***

## KNN

The KNN classifier produced strong predictive accuracy, with performance remaining around 82.5% accuracy across several values of K. The best financial outcome occurred when K = 3, producing a total cost of -137,245.

| K | Accuracy | Total Cost |
|---|---|---|
| 3 | 0.8268 | -137,245 |
| 5 | 0.8298 | -202,195 |
| 7 | 0.8294 | -297,765 |
| 10 | 0.8311 | -712,370 |
| 25 | 0.8275 | -609,800 |
| 50 | 0.8224 | -1,012,620 |
| 100 | 0.8151 | -1,286,945 |

While KNN achieved good classification accuracy, the model still produced negative financial outcomes. As K increased, the model became more conservative in predicting cancellations, which unsuprisingly lead to a higher number of missed cancellations (false negatives). These are incredibly costly for BRI.

## Logistic Regression 

Logistic regression produced stable performance across all tested regularization values. Accuracy remained around 79.5%, and changes to the regularization parameter C had minimal impact on overall accuracy and total cost.

| C | Accuracy | Total Cost |
|---|---|---|
| 0.01 | 0.7949 | -1,766,415 |
| 0.1 | 0.7952 | -1,745,240 |
| 1 | 0.7952 | -1,740,160 |
| 10 | 0.7954 | -1,738,115 |
| 100 | 0.7954 | -1,737,460 |

Although logistic regression performed reasonably in terms of classification accuracy, it had the worst financial outcomes among the three model types across all configurations. Logistic regression is too simplistic for this problem.

## Decision Tree

The Decision Tree classifier showed the most significant improvement as model complexity increased. Shallow trees performed poorly, with lots of false negatives and high costs. As the maximum depth increased however, the model was able to capture the more complex relationships present within this dataset.

| Max Depth | Accuracy | Total Cost |
|---|---|---|
| 3 | 0.7505 | -4,155,995 |
| 5 | 0.7911 | -2,088,800 |
| 10 | 0.8394 | -283,515 |
| 25 | 0.8470 | 257,115 |
| None | 0.8447 | 284,485 |

The best configuration occurred when there was no max depth. This achieved the highest accuracy and produced a positive financial outcome. This configuration of the decision tree model would generate net savings for the company.

## Overall

| Model | Best Parameter | Accuracy | Total Cost |
|------|------|------|------|
| KNN | K = 3 | 0.8268 | -137,245 |
| Logistic Regression | C = 100 | 0.7954 | -1,737,460 |
| Decision Tree | max_depth = none | **0.8447** | **284,485** |

From the table above, the Decision Tree classifier outperforms the other models in both predictive performance and financial outcome.

***
## Section: 7.2 - Choose a Best Classifier
- Choose one of the three classifiers as best and explain why.
***

Based on both classification accuracy and financial impact, the Decision Tree classifier is the best model for predicting cancellations. It had the highest accuracy and was the only model capable of producing a positive financial outcome for BRI.

The results suggest that the relationship between reservation characteristics and cancellation behavior is quite complex and nonlinear. Decision trees are well suited for capturing these type of patterns, which explains they had such great performance compared to the much simpler logisticc regression model and the distance-based KNN classifier.

If implemented in practice, the Decision Tree model could allow the company to more effectively identify reservations that are likely to cancel and selectively apply their new incentive program. This targeted intervention strategy would reduce the unnecessary incentive costs of the current plan, while simultaneously 35% of cancellations. BRI would become more profitable by this decision.

***
# Section: 8 - Conclusions
- Write a paragraph on what you discovered or learned from this homework.
- What are your overall conclusions about the data?
- What did you learn? What would you explore further with additional data, time or resources. What might "future research" require to gain deeper insight? 
***

One key takeaway from this assignment is that the model with the highest classification accuracy is not necessarily the best model financially. While KNN and logistic regression produced very reasonable predictive accuracy, they still resulted in negative financial outcomes due to missed cancellations (false negatives). The decision tree classifier performed the best overall, achieving the highest accuracy and the only positive financial benefit, suggesting that the relationship between reservation characteristics and cancellations is complex. This means the dataset is more complex than I originally perceived when reviewing it by hand, and more than sufficient for analyzing this cancellation problem. I would not have predicted the decision tree model to be the best performing before starting this assignment. If there was more time and reousrces available, I could further explore by trying other models like SVM to compare with the already written models. It's possible these models might perform better. I could further tune model parameters as well, to try to squeeze the best possible peformance out of every model, and perhaps gain deeper insight into the dataset and have more analysis to provide from a business perspective.

***
### END-OF-SUBMISSION
***